Software/Development/SemesterProject/App/referenceHandler.c

```diff
@@ -0,0 +1,317 @@
+ */
+static GUISignalsTag GUISignals;
+static float speedReference, torqueReference;
+static float deltaSpeed = 5, deltaTorque = 0.1;
+static int undampedSpeedReference = 0;
+static int referenceSourceChanged = 0;
+
+
+static referenceSourceTag referenceSource = interfacePCB;
+static enum referenceTypeTag referenceType = torqueControl;
+const int SPEED_STEP_FROM_BUTTON_RPM = 60 * SPEED_STEP_FROM_BUTTON_KPH / 3.6 * GEAR_RATIO / RADIUS_WHEELS;
+
+
+void handleReferences(void)
+{
+    GUISignals = getGUISignals();
+    decideReferenceSource();
+    if (readSystemState() == RUNNING)
+    {
+        calculateReference();
+        decideReferenceType();
+    }
+}
+
+/*
+ * This FSM decides what reference is going to be used for setting a torque
+ * reference. The options are Interface PCB, Pedal or Graphical user interface.
+ */
+void decideReferenceSource(void)
+{
+    switch (referenceSource)
+        {
+        case interfacePCB:
+        {
+            setTorqueReferenceSliderLED(ON); // Necessary for system startup
+            if (referenceSourceIsChanged())
+            {
+                setTorqueReferenceSliderLED(OFF);
+                referenceSource = goKart;
+                setTorqueReferencePedalLED(ON);
+            }
+        }break;
+        case goKart:
+        {
+            if (referenceSourceIsChanged())
+            {
+                setTorqueReferencePedalLED(OFF);
+                referenceSource = GUI;
+                setTorqueReferenceUARTLED(ON);
+            }
+        }break;
+        case GUI:
+        {
+            if (referenceSourceIsChanged())
+            {
+                setTorqueReferenceUARTLED(OFF);
+                referenceSource = interfacePCB;
+                setTorqueReferenceSliderLED(ON);
+            }
+        }break;
+        }
+}
+
+/*
+ * This FSM decides what type of control is used for controlling the motor: Torque
+ * control or cruise control. If the system is being torque controlled and the user
+ * requests cruise control, the instantaneous rotor speed will be acquired and will
+ * be used as the reference for the control algorithm.
+ */
+void decideReferenceType(void)
+{
+    setCruiseControlLED(OFF);
+    switch (referenceType)
+    {
+    case torqueControl:
+    {
+        if (referenceTypeIsChanged())
+        {
+            undampedSpeedReference = 1000;// THIS MUST BE ADDED WHEN THE INTERFACE FROM CONTROL IS CREATED getCurrentRotorSpeed();
+            referenceType = cruiseControl;
+        }
+    }break;
+    case cruiseControl:
+    {
+        if (referenceTypeIsChanged())
+        {
+            referenceType = torqueControl;
+            restartSpeedReference();
+        }
+        setCruiseControlLED(ON);
+    }break;
+    }
+}
+
+void calculateReference(void)
+{
+    calculateTorqueReference();
+    calculateSpeedReference();
+}
+
+void calculateTorqueReference(void)
+{
+    float torqueReferenceBeforeLimit = 0, torqueReferenceBeforeSaturation = 0;
```

Comment:
WHy init it to 0? no big deal

Reply:
Not a big deal, just for seeing a nice '0' when debugging. Otherwise it was init to some random 46468468454e-48.